### DISCUSSION TITLE
DeepLabCE and ignore_label / ignore_index

### DISCUSSION THREAD

**Discussion by Giaco90** [facebookresearch/detectron2#3876] [Q&A]
Hello all,
reading the code of [DeepLabCE](https://github.com/facebookresearch/detectron2/blob/38628d1f0d25f1186985fa2d630d5db4ad162ddb/projects/DeepLab/deeplab/loss.py), I found a possible problem in how the forward method is implemented.
The used criterion is CrossEntropyLoss and is set up with ignore_index parameter and reduction="none"; according to [CrossEntropyLoss doc](https://pytorch.org/docs/1.9.1/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss):

> ignore_index (int, optional) – Specifies a target value that is ignored and does not contribute to the input gradient. When size_average is True, the loss is averaged over non-ignored targets.

then, ignore_index is used to:
- modify the loss in order to ignore specific values (I think something like zeroing specific positions)
- ignore values during the reduction (the doc says during size_average, but it has been deprecated so i think it does the same with reduction set to sum or mean).

The problem is that the forward of DeepLabCE applies the CrossEntropyLoss without the reduction and manually applies the mean after weighting and topk selection. Then, the first point is in someway applied, but the second is not.

Is this correct? Should the mean be applied considering the ignore_label / ignore_index ?

Thanks,
Giacomo
